WillTrem/UInnovate
# #91 Use Case API Complete test suite with SoapUI

This PR marks the completion of #91, and with it comes the final transition away from the Postman test suite. I now have fully set up the base for all use case API testing, with a substantial amount of the core URL functionality being verified in here. This is done with the help of a new testing tool, SoapUI.
![image](https://github.com/WillTrem/UInnovate/assets/61199167/af7fe621-d731-4bc2-a7bb-e2848fb3eb98)

Its a fully UI based open-source API testing program, so it seemed like a good alternative to Postman (it took an enormous amount of effort to learn properly though).

Here is the essence of the created functionality:
- Both of app_rentals and app_service_support have various tailored test cases that cover much of the functionality used from our URLs, with standard Get, Post, Patch, Put and Delete requests, as well as pagination, ordering, filtering and table joins.

- There is also a Universal test case, which includes a lengthy script which in essence should cover the basic get requests of *every future use case we may have*, as long as it follows the app_* schema syntax

- There are load tests on all of the aformentioned get requests (I opted not to load test the other types of requests, because they would alter the database multiple times a second and would be difficult to track from a table view side). These include 3 different strategies: simple constant load, burst load, and thread ramp up load

- There is a beginning to some security tests for the url accessing and login, but there is still work to be done to secure the database, so this would be difficult to design prematurely

## Steps for testing

1. Navaigate to https://www.soapui.org/downloads/soapui/, and download "SoapUI Open Source"

2. Use the installer you downloaded to install SoapUI. You can choose all the defaults in the installation wizard, but I recommend also allowing a desktop shortcut
 
3. Open SoapUI once it is finished, and after closing the opening pop-up, click on "File" in the top left, and then "Import Project"
![image](https://github.com/WillTrem/UInnovate/assets/61199167/d3ef057d-6afb-4f3e-adc5-cbbf16533199)

4. Navigate to *\UInnovate\api_testing, wherever that is on your computer, and select the UInnovate-soapui-project.xml file
![image](https://github.com/WillTrem/UInnovate/assets/61199167/0f02ad20-d1f8-46ef-9b77-a6c03630d925)

5. This should load up the test suite and API calls in the navigator, you can open the dropdown for the testsuite to see all the test cases. Simply double click on a test case to have a little window pop up with all the test steps, if you press the play button in that window it will run them all.
![image](https://github.com/WillTrem/UInnovate/assets/61199167/f9755d4d-023e-409f-be1f-53f7f460b50d)
You can double click on individual test steps to see more details, such as each endpoint and the assertions it runs. Authentication has also been implemented, so you can turn off login bypass and it should still work (test that out too please)

6. The test cases marked "GET ONLY" are for the load tests, to run these you need to open the drop down on those test cases, and then the sub drop down called "Load tests". Once you see these, you can double click any one of them to open the load test panel, and click the green play button to run them. Note that these may take quite a bit of your cpu, so it might be slower on some computers (I tried to be conservative with thread count here, but you've been warned). There are also 2 buttons next to the stop button for graphs, if you want to plot the overall results of the load tests. 
![image](https://github.com/WillTrem/UInnovate/assets/61199167/040c4b5d-0f65-41fa-9667-a35a0a11edf4)

That should be about it! There might be some more ways to flesh this out in the final iterations, such as adding this to the CI/CD pipeline and doing security tests once thats been handled database side, but API testing at its core is completely handled with these changes.
